realtime/src/main/java/com/c8y/ms/templates/realtime/controller/NotificationController.java

```diff
@@ -0,0 +1,166 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Controller;
+import org.svenson.JSON;
+import org.svenson.JSONParser;
+
+@Controller
+public class NotificationController {
+    private static final Logger LOG = LoggerFactory.getLogger(NotificationController.class);
+    /**
+     * device id is stored in the application.properties as a property
+     */
+    @Value("${device.id}")
+    private String deviceId;
+
+    @Autowired
+    private PlatformImpl platform;
+
+    @Autowired
+    private ContextService<MicroserviceCredentials> contextService;
+
+    private MicroserviceCredentials microserviceCredentials;
+
+    @Autowired
+    private MicroserviceSubscriptionsService subscriptions;
+
+    @Autowired
+    private OperationService operationService;
+
+    private Subscription<String> subscription;
+
+    private Subscription<String> measurementSubscription;
+
+    private Subscription<GId> operationSubscription;
+
+    private final JSONParser jsonParser = JSONBase.getJSONParser();
+
+    private final JSON json = JSON.defaultJSON();
+
+    /**
+     * This method will be triggered during the initial start of the Microservice
+     * when all subscriptions are queried and during runtime when a tenant
+     * subscribes for this particular Microservice.
+     *
+     * @param event contains information about the tenant which subscribed for this Microservice
+     */
+    @EventListener
+    private void onMicroserviceSubscriptionAddedEvent(final MicroserviceSubscriptionAddedEvent event) {
+        microserviceCredentials = contextService.getContext();
+
+        // execute business logic in the context of a particular tenant
+        subscriptions.runForTenant(event.getCredentials().getTenant(), () -> {
+            registerForDeviceUpdates();
+            registerForMeasurementUpdates();
+            registerForOperations();
+        });
+    }
+
+    /**
+     * Subscribe for device updates. In this case the subscription is done for a
+     * specific device but using the wildcard character * you can subscribe for
+     * updates of all managed objects. For subscription the built-in class
+     * InventoryRealtimeDeleteAwareNotificationsSubscriber is being used.
+     */
+    private void registerForDeviceUpdates() {
+        InventoryRealtimeDeleteAwareNotificationsSubscriber subscriber = new InventoryRealtimeDeleteAwareNotificationsSubscriber(platform);
+        this.subscription = subscriber.subscribe(deviceId, new SubscriptionListener<>() {
+            @Override
+            public void onNotification(Subscription<String> arg0, ManagedObjectDeleteAwareNotification arg1) {
+                final ManagedObjectRepresentation mo = jsonParser.parse(ManagedObjectRepresentation.class, json.forValue(arg1.getData()));
```

Comment:
Be aware that this will throw an error if device gets deleted, you use InventoryRealtimeDeleteAwareNotificationSubscriber, in case of delete you have to handle arg1.getData() different, it is a String instead of ManagedObjectRepresentation.

```suggestion
if("DELETE".equals(notification.getRealtimeAction())) {
            	String managedObjectId = (String) notification.getData();
            	logger.info("ManagedObject with ID {} removed.", managedObjectId);
}
else {
                final ManagedObjectRepresentation mo = jsonParser.parse(ManagedObjectRepresentation.class, json.forValue(arg1.getData()));
                LOG.info("on device update received: {}", mo.toJSON());

                }
```